From 933cbab3ba056f9be25715f4e1148746611b4fb6 Mon Sep 17 00:00:00 2001
From: Mohan Kumar G <quic_mkumarg@quicinc.com>
Date: Mon, 24 Jun 2024 18:55:26 +0530
Subject: [PATCH] hostapd: remove duplicate puncturing bitmap

Currently, the puncturing bitmap is passed two times,
one in frequency parameters and other in wpa_driver_ap_params.

The bitmap in frequency paramters is used to update the punctured
channels. Hence remove the puncturing bitmap in
wpa_driver_ap_params. Similarly, use the frequency parameters
for mesh puncturing bitmap.

And, update correct puncturing bitmap during channel switch.


Signed-off-by: Mohan Kumar G <quic_mkumarg@quicinc.com>
---
 src/drivers/driver_nl80211.c | 7 -------
 1 file changed, 7 deletions(-)

--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -5560,13 +5560,6 @@ static int wpa_driver_nl80211_set_ap(voi
 		if (nla_put_u8(msg, NL80211_ATTR_SET_CRITICAL_UPDATE, critical_update))
 			goto fail;
 	}
-	if (params->punct_bitmap) {
-		wpa_printf(MSG_DEBUG, "nl80211: Puncturing bitmap=0x%04x",
-			   params->punct_bitmap);
-		if (nla_put_u32(msg, NL80211_ATTR_PUNCT_BITMAP,
-				params->punct_bitmap))
-			goto fail;
-	}
 
 #ifdef CONFIG_DRIVER_NL80211_QCA
 	if (cmd == NL80211_CMD_NEW_BEACON && params->allowed_freqs)
@@ -12468,13 +12461,6 @@ static int nl80211_join_mesh(struct i802
 	      nla_put_u32(msg, NL80211_ATTR_BEACON_TX_MODE, params->beacon_tx_mode)))
 		goto fail;
 
-	if (params->punct_bitmap) {
-		wpa_printf(MSG_DEBUG, "nl80211: Puncturing bitmap=0x%04x",
-			   params->punct_bitmap);
-		if (nla_put_u32(msg, NL80211_ATTR_PUNCT_BITMAP,
-				params->punct_bitmap))
-			goto fail;
-	}
 	wpa_printf(MSG_DEBUG, "  * flags=%08X", params->flags);
 
 	if (params->handle_dfs && nla_put_flag(msg, NL80211_ATTR_HANDLE_DFS))
--- a/wpa_supplicant/mesh.c
+++ b/wpa_supplicant/mesh.c
@@ -283,7 +283,7 @@ static int wpas_mesh_complete(struct wpa
 		wpa_s->mgmt_group_cipher = wpa_s->mesh_rsn->mgmt_group_cipher;
 	}
 
-	params->punct_bitmap = ifmsh->conf->punct_bitmap;
+	params->freq.punct_bitmap = ifmsh->conf->punct_bitmap;
 	params->ies = ifmsh->mconf->rsn_ie;
 	params->ie_len = ifmsh->mconf->rsn_ie_len;
 	params->basic_rates = ifmsh->basic_rates;
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -1385,6 +1385,7 @@ int hostapd_parse_csa_settings(const cha
 	SET_CSA_SETTING_EXT(punct_bitmap);
 	SET_CSA_SETTING(bandwidth_device);
 	SET_CSA_SETTING(center_freq_device);
+	settings->freq_params.punct_bitmap = settings->punct_bitmap;
 	settings->freq_params.ht_enabled = !!os_strstr(pos, " ht");
 	settings->freq_params.vht_enabled = !!os_strstr(pos, " vht");
 	settings->freq_params.he_enabled = !!os_strstr(pos, " he");
